frontend/src/stimulus/openproject-stimulus-application.ts

```diff
@@ -0,0 +1,20 @@
+import { ControllerConstructor } from '@hotwired/stimulus/dist/types/core/controller';
+import { Application } from '@hotwired/stimulus';
+
+export class OpenProjectStimulusApplication extends Application {
+  static controllers:{ [name:string]:ControllerConstructor } = {};
+
+  static register(name:string, controller:ControllerConstructor) {
+    this.controllers[name] = controller;
+  }
+
+  async start():Promise<void> {
+    Object
+      .keys(OpenProjectStimulusApplication.controllers)
+      .forEach((name:string) => {
+        this.register(name, OpenProjectStimulusApplication.controllers[name]);
+      });
+
+    await super.start();
+  }
```

Comment:
Two minor suggestions:

- using a `Map` is a more modern approach and makes it easier to iterate: maps also remember the initial insertion order.
- naming: after thinking it over, I think it would be better to call the registration method something other than `register`. `preregister` is a suggestion - to make it clear that these controllers will be registered before the Stimulus application is started.

```suggestion
  static controllers = new Map<string, ControllerConstructor>();

  static preregister(name:string, controller:ControllerConstructor) {
    this.controllers.set(name, controller);
  }

  async start():Promise<void> {
    this.preregisteredControllers.forEach((controller, name) => {
      this.register(name, controller);
    });

    await super.start();
  }

  get preregisteredControllers() {
    return OpenProjectStimulusApplication.controllers;
  }
```

Reply:
I like that better too, thanks! 🙏